From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alex Betts <alex.betts@gmail.com>
Date: Tue, 27 Dec 2022 12:10:53 -0500
Subject: bbg: Generate 'v8-gn.h' without using it in V8

This patch adds a flag,
`v8_generate_external_defines_header_for_embedder_only`, to V8's
`BUILD.gn`.

When this flag is `true`, the header `v8-gn.h` will be generated by the
build but not used when building V8 itself, because doing so generates
'undefined symbol' errors when linking `mksnapshot` (in particular,
translation unit generated for 'remembered-set.cc' will be empty).

diff --git a/BUILD.gn b/BUILD.gn
index c3393a116390d5321da578110c9d3ab227126351..81e2f8c54b9f6de76e610d5bdc252858da63bebf 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -330,6 +330,11 @@ declare_args() {
   # header files are placed in a separate header file v8-gn.h.
   v8_generate_external_defines_header = false
 
+  # If enabled then macro definitions that are used in externally visible
+  # header files are placed in a separate header file v8-gn.h, but, V8 itself
+  # will not use it
+  v8_generate_external_defines_header_for_embedder_only = false
+
   # Experimental feature for tracking constness of properties in non-global
   # dictionaries. Enabling this also always keeps prototypes in dict mode,
   # meaning that they are not switched to fast mode.
@@ -2553,6 +2558,8 @@ v8_header_set("v8_config_headers") {
   if (v8_generate_external_defines_header) {
     sources += [ "$target_gen_dir/include/v8-gn.h" ]
     deps += [ ":gen_v8_gn" ]
+  } else if (v8_generate_external_defines_header_for_embedder_only) {
+    deps += [ ":gen_v8_gn" ]
   }
 }
 
@@ -2630,7 +2637,8 @@ v8_header_set("v8_headers") {
   ]
 }
 
-if (v8_generate_external_defines_header) {
+if (v8_generate_external_defines_header ||
+    v8_generate_external_defines_header_for_embedder_only) {
   action("gen_v8_gn") {
     visibility = [ ":*" ]
 
